agent/src/main/java/com/appland/appmap/transform/annotations/Hook.java

```diff
@@ -159,48 +139,49 @@ public static void apply(CtBehavior targetBehavior, List<HookSite> hookSites) {
       } else {
         invocations[index] += hookSite.getHookInvocation();
       }
-    }
+      String uniqueKey = hookSite.getUniqueKey();
+      if (!uniqueKey.isEmpty()) {
+        if (!uniqueKeys.contains(uniqueKey)) {
+          uniqueLocks.append("com.appland.appmap.process.ThreadLock.current().lockUnique(\"")
+              .append(uniqueKey).append("\");");
+          uniqueKeys.add(uniqueKey);
+        }
+      }
 
-    final String uniqueLocks = hookSites
-        .stream()
-        .map(HookSite::getUniqueKey)
-        .filter(uk -> !uk.isEmpty())
-        .distinct()
-        .map(uniqueKey -> (""
-            + "com.appland.appmap.process.ThreadLock.current().lockUnique(\""
-            + uniqueKey
-            + "\");"))
-        .collect(Collectors.joining("\n"));
+    }
 
     try {
+      String beforeSrcBlock = beforeSrcBlock(uniqueLocks.toString(),
+          invocations[MethodEvent.METHOD_INVOCATION.getIndex()]);
+      logger.trace("{}: beforeSrcBlock:\n{}", targetBehavior::getName, beforeSrcBlock::toString);
       targetBehavior.insertBefore(
-          beforeSrcBlock(uniqueLocks, invocations[MethodEvent.METHOD_INVOCATION.getIndex()])
-      );
+          beforeSrcBlock);
+
+      String afterSrcBlock = afterSrcBlock(invocations[MethodEvent.METHOD_RETURN.getIndex()]);
+      logger.trace("{}: afterSrcBlock:\n{}", targetBehavior::getName, afterSrcBlock::toString);
 
       targetBehavior.insertAfter(
-          afterSrcBlock(invocations[MethodEvent.METHOD_RETURN.getIndex()]));
+          afterSrcBlock);
 
       ClassPool cp = AppMapClassPool.get();
+      String exitEarlyCatchSrc = "{com.appland.appmap.process.ThreadLock.current().exit();return;}";
```

Comment:
Is there a way to precompile this and the other blocks so that we don't recompile the same code over and over for every method?

Reply:
No, javassist doesn't offer a way to do this. :(

Reply:
Perhaps this could be done with ByteBuddy later, if it turns out to be a performance drain.

Reply:
Yeah, it would definitely be better to do all instrumentation using Byte Buddy. That was the approach I took initially, but it turned out to be an even heavier lift.

Reply:
We can get there gradually I suppose.